Source/Core/VideoCommon/Resources/MaterialResource.cpp

```diff
@@ -0,0 +1,391 @@
+                 if (resource)
+                   resource->MarkAsActive();
+               }},
+               texture_resource);
+  }
+  if (m_current_data->m_shader_resource)
+    m_current_data->m_shader_resource->MarkAsActive();
+  if (m_current_data->m_next_material)
+    m_current_data->m_next_material->MarkAsActive();
+}
+
+void MaterialResource::MarkAsPending()
+{
+  m_resource_context.asset_cache->MarkAssetPending(m_material_asset);
+}
+
+void MaterialResource::CreateTextureData(Data* data)
+{
+  ShaderCode preprocessor_settings;
+
+  const auto& material_data = *data->m_material_data;
+  data->m_texture_like_data.clear();
+  data->m_texture_like_resources.clear();
+  data->m_texture_like_references.clear();
+  const u32 custom_sampler_index_offset = 8;
+  for (u32 i = 0; i < static_cast<u32>(material_data.textures.size()); i++)
+  {
+    const auto& texture_and_sampler = material_data.textures[i];
+    data->m_texture_like_references.push_back(TextureLikeReference{});
+
+    TextureAndSamplerResource* value = nullptr;
+    data->m_texture_like_resources.push_back(value);
+    data->m_texture_like_data.push_back(std::shared_ptr<TextureAndSamplerResource::Data>{});
+
+    auto& texture_like_reference = data->m_texture_like_references[i];
+
+    if (texture_and_sampler.asset == "")
+    {
+      preprocessor_settings.Write("#define HAS_SAMPLER_{} 0\n", i);
+
+      // For an invalid asset, force the sampler to use the default sampler
+      texture_like_reference.sampler_origin =
+          VideoCommon::TextureSamplerValue::SamplerOrigin::Asset;
+      texture_like_reference.texture_hash = "";
+    }
+    else
+    {
+      preprocessor_settings.Write("#define HAS_SAMPLER_{} 1\n", i);
+
+      texture_like_reference.sampler_origin = texture_and_sampler.sampler_origin;
+      texture_like_reference.texture_hash = texture_and_sampler.texture_hash;
+    }
+
+    texture_like_reference.sampler_index = i + custom_sampler_index_offset;
+    texture_like_reference.texture = nullptr;
+  }
+
+  data->m_preprocessor_settings = preprocessor_settings.GetBuffer();
+}
+
+void MaterialResource::SetShaderKey(Data* data, GXPipelineUid* uid)
+{
+  XXH3_state_t shader_key_hash;
+  XXH3_INITSTATE(&shader_key_hash);
+  XXH3_64bits_reset_withSeed(&shader_key_hash, static_cast<XXH64_hash_t>(1));
+
+  UpdateHashWithPipeline(*uid, &shader_key_hash);
+  XXH3_64bits_update(&shader_key_hash, data->m_preprocessor_settings.c_str(),
+                     data->m_preprocessor_settings.size());
+
+  data->m_shader_key = XXH3_64bits_digest(&shader_key_hash);
+}
+
+void MaterialResource::WriteUniforms(Data* data)
+{
+  // Calculate the size in memory of the buffer
+  std::size_t max_uniformdata_size = 0;
+  for (const auto& property : data->m_material_data->properties)
+  {
+    max_uniformdata_size += VideoCommon::MaterialProperty::GetMemorySize(property);
+  }
+  data->m_uniform_data.resize(max_uniformdata_size);
```

Comment:
You could make `m_uniform_data` a `Common::UniqueBuffer<u8>` to avoid the unnecessary zero initialization here, assuming this isn't relying on unfilled data being zereoed or something.